runtime/common/src/migrations/loans.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright 2024 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::PoolNAV;
+use frame_support::{
+	pallet_prelude::{Decode, Encode, StorageVersion},
+	traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+	weights::Weight,
+};
+use pallet_loans::{pallet::Pallet as Loans, Config};
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_runtime::TryRuntimeError;
+use sp_std::vec::Vec;
+
+const LOG_PREFIX: &str = "LoansMigrationToV4:";
+
+mod v3 {
+	use frame_support::{
+		pallet_prelude::{OptionQuery, ValueQuery},
+		storage_alias, Blake2_128Concat, BoundedVec,
+	};
+	pub use pallet_loans::entities::{
+		loans::v3::ActiveLoan, pricing::external::v3::CreatedLoan as CreatedLoanStruct,
+	};
+	use pallet_loans::Config;
+
+	use super::Loans;
+
+	pub type ActiveLoansVec<T> =
+		BoundedVec<(<T as Config>::LoanId, ActiveLoan<T>), <T as Config>::MaxActiveLoansPerPool>;
+
+	#[storage_alias]
+	pub type ActiveLoans<T: Config> = StorageMap<
+		Loans<T>,
+		Blake2_128Concat,
+		<T as Config>::PoolId,
+		ActiveLoansVec<T>,
+		ValueQuery,
+	>;
+
+	#[storage_alias]
+	pub type CreatedLoan<T: Config> = StorageDoubleMap<
```

Comment:
Just curious, are we sure we do not need to migrate `ClosedLoan`? Will the expected decoded vector with that empty value match the current structure?

Reply:
What I wanted to say: We only need to migrate if this storage is not empty. Right now, it is empty and it seems reasonable we can keep this state until upgrading Mainnet.

However, then I checked Demo and it has one entry. So yes, I will write a migration!

Reply:
Ah ok, I didn't know it was empty in centrifuge

Reply:
`ClosedLoan` migration added here: [833b9d1](https://github.com/centrifuge/centrifuge-chain/pull/1838/commits/833b9d125a15b4bb3328f60fe17f51ffebfe30d5)